### ISSUE TITLE
Refactor: Vectorize Rule Evaluation in RuleLayer to Mitigate Combinatorial Explosion

### ISSUE THREAD

**Issue by dcruzf** [dcruzf/highFIS#53]
The current implementation of `RuleLayer` relies on a sequential Python `for` loop to compute
the firing strengths of the antecedent rules during the `forward()` pass. For high-dimensional
TSK systems, the rule base grows exponentially with the number of inputs \u2014 a phenomenon known
as **Combinatorial Explosion**:

| Inputs (D) | MFs per input | Total Rules |
|:---:|:---:|:---:|
| 5 | 3 | 243 |
| 10 | 3 | 59,049 |
| 15 | 3 | 14,348,907 |
| 20 | 3 | 3,486,784,401 |

Iterating over tens of thousands of rules sequentially inside `forward()` breaks PyTorch's
ability to parallelize work on the GPU, creating a severe CPU/GPU synchronization bottleneck
and drastically reducing training throughput.

---

### Standard TSK: Product T-norm

In a standard TSK model, the firing strength $w_r$ of rule $r$ across $D$ dimensions is:

$$w_r = \prod_{d=1}^{D} \mu_{r,d}(x_d)$$

For high-dimensional spaces, multiplying many values in $[0, 1]$ causes $w_r \to 0$,
leading to **vanishing gradients** during backpropagation.

### HTSK: Geometric Mean

The **HTSK** architecture addresses this by normalizing with the geometric mean:

$$w_r = \left(\prod_{d=1}^{D} \mu_{r,d}(x_d)\right)^{1/D}$$

### Numerical Stability via Log-Space

To ensure stability and enable backpropagation, the operation is computed in log-space:

$$\log w_r = \frac{1}{D} \sum_{d=1}^{D} \log \mu_{r,d}(x_d)$$
$$w_r = \exp\left(\frac{1}{D} \sum_{d=1}^{D} \log \mu_{r,d}(x_d)\right)$$

Currently, the library computes this one rule at a time. To fully leverage hardware
accelerators, we must lift the entire computation into **tensor algebra**, computing
$\mathbf{W}$ (the matrix of all $w_r$ for all samples in the batch) in a single parallel operation.

---

### Core Idea

Remove the Python loop from `RuleLayer.forward()`. Instead:

1. **At `__init__` time:** Pre-compute a static index matrix `rule_indices` of shape
   `(n_rules, n_inputs)` that maps every rule to its absolute flat index in the
   concatenated membership tensor.
2. **At `forward()` time:** Use PyTorch advanced indexing to fetch all antecedent
   membership values simultaneously, producing a single tensor of shape
   `(batch_size, n_rules, n_inputs)`, then reduce over `dim=2` with the chosen T-norm.

### Implementation \u2014 `highfis/layers.py`

```python
import torch
from torch import Tensor, nn


class RuleLayer(nn.Module):
    def __init__(
        self,
        input_names: list[str],
        mf_per_input: list[int],
        rules: list[tuple[int, ...]],
    ):
        super().__init__()
        self.input_names = input_names
        self.mf_per_input = mf_per_input
        self.rules = rules
        self.n_rules = len(rules)
        self.n_inputs = len(input_names)

        # Pre-compute absolute indices once to avoid runtime loop overhead.
        self._register_vectorized_indices()

    # ------------------------------------------------------------------
    # Initialisation helpers
    # ------------------------------------------------------------------

    def _register_vectorized_indices(self) -> None:
        """Convert relative per-input MF indices into absolute flat indices.

        Given mf_per_input = [3, 2], the offsets are [0, 3].
        A rule (1, 0) maps to absolute indices [1+0, 0+3] = [1, 3].
        """
        offsets: list[int] = []
        current = 0
        for n_mfs in self.mf_per_input:
            offsets.append(current)
            current += n_mfs

        abs_rules = [
            [mf_idx + offsets[i] for i, mf_idx in enumerate(rule)]
            for rule in self.rules
        ]

        # register
